Add mouse support

omahlama opened the pull request:
This library offered exactly what I needed, i.e. complete control over the rendering of cards, including the in-between states. I needed mouse support in addition to touch, so I added it in this PR.

Notable things:
- mouse move and up events listeners are added to document, so that user can drag outside the carousel and things work normally
- mouse move and up handlers do nothing, unless we are dragging
- being able to drag outside the carousel exposed a bug where the state was empty until the next start event - fixed by adding `modCursor` on spring rest

Let me know if you need anything to get this merged, we're now building our app on top of my fork, but it would be great to get to get everything upstream.

xiaody commented:
Mouse support was considered when I came up the idea to build this component. But I thought that people rarely use mouse to **drag** a carousel on desktop. The "arrow buttons" pattern is much more common, something like what https://www.amazon.com/ and https://www.taobao.com/ use.

So I decided to support only touch events and expose some methods like `prev()` and `next()` so the users of this component can easily build up the "arrow buttons" pattern if they need.

I'm wondering what does your app's UX looks like? Why do you want your users to drag a carousel using a mouse? You know, the mouse drag gesture feels harder than clicking a button on desktop, especially when the user is using some non-mouse input device like touchpads.

omahlama commented:
I agree that dragging with a mouse is not ideal, the correct functionality would be horizontal scrolling, but adding horizontal scroll bars to a webpage just looks horrible. We'll add buttons that use the `prev` and `next`, but the requirement from designers is to also allow dragging with the mouse. At least with the big hero banner carousel, flicking to the next image feels quite easy and natural with the macbook touchpad.

omahlama commented:
For example the other library I evaluated (react-slick with almost 4k github stars) allows dragging with both mouse and touch - I just couldn't use it because it doesn't allow the same kind of control over the rendering of cards depending on the distance that has been dragged. See their demo page http://neostack.com/opensource/react-slick

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,22 +55,37 @@ class TouchCarousel extends React.PureComponent {
 
   componentWillUnmount () {
+    document.removeEventListener('mousemove', this.onMouseMove)
+    document.removeEventListener('mouseup', this.onMouseUp)
```
Maybe name like `onDocumentMouseup`? To distinguish from other handlers on the container.

omahlama replied:
Good point, will fix.

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,22 +55,37 @@ class TouchCarousel extends React.PureComponent {
+    this.onStart(e.changedTouches.length, getTouchId(e), e)
+  }
+
+  onStart = (changeCount, touchId, e) => {
```
What about keeping the original `onTouchStart`, `onTouchMove`, `onTouchEndOrCancel` untouched and mock up some events in the mouseEvent handlers and pass the mocked events to the touchEvent handlers? Maybe it's more clear and we keep touch support as first class.

omahlama replied:
To be honest I did it this way because I didn't want to mock anything - I feel that explicit parameters are clearer than mocking touch events. If you look at https://github.com/xiaody/react-touch-carousel/pull/4/commits/6f5d76b42e6bf5004ca84fcd343835a7847166bd, you'll see that the changes are quite minimal, `onStart` is practically the same as the old `onTouchStart`, with 2 lines of code moved to the new `onTouchStart`

xiaody replied:
Yes, I see the changes are minimal.
But for me, the `onStart(changeCount, touchId, e)` signature looks quite confusing:

- changeCount: why does it need a count?
- touchId: umm ok this makes sense
- e: oh finally you need still the `e` so why don't you just calculate the count and id by yourself?

And when look into the function body, it's still all about "touch".
If you stick to this way, you'd better rename all the things from "touch" to say "pointer".

xiaody replied:
I'd prefer mocking touch events anyway.

xiaody commented on src/TouchCarousel.js:
```diff
@@ -198,6 +236,7 @@ class TouchCarousel extends React.PureComponent {
       modIndex += this.props.cardCount
     }
     this.props.onRest(index, modIndex, cursor, this.state)
+    this.modCursor()
```
> being able to drag outside the carousel exposed a bug where the state was empty until the next start event...

I don't really get this. Can you explain more?
The moding was always done in touchStart event.

omahlama replied:
With the mouse (and mouseMove and mouseUp events hooked to `document`) you can start dragging from the carousel and drag very far to the left or right, further than what padCards are filled. This means that the user sees empty state until the next time that `modCursor` is called. This doesn't happen with touch, because you can only drag to the end of the TouchCarousel, so the padCards are always enough.

Without this line, the user would be stuck in the empty state because `modCursor` was only called the next time they start drag. With this line, when the user stops dragging and the spring is at rest, we always call `modCursor` and the user sees the cards.

It would be event better, if the user never saw the empty state, but calling `modCursor` in the middle of user drag caused problems (drag got stuck), so this was a good enough solution.

omahlama replied:
You can test this for yourself at http://localhost:5000/boring.html?loop&autoplay, drag with the mouse very fast and far.

xiaody replied:
The `padCount` are designed for this. PadCards should always fill the carousel area no matter how far the user drags. If you see empty state, you need a bigger `padCount`.

Also, I think `modCursor()` in the spring rest event doesn't fully resolve your problem: the user has already seen the empty state before releasing the mouse.

omahlama replied:
I increased the padCount in my for this reason, but still doing the `modCursor` here is good as defensive programming: if the user can get to the empty state, at least they don't get stuck.

xiaody replied:
Defensive programming is for things that you cannot foresee, like drunk users' input and network errors. But in our case, the component's state is under fully control. We design and know exactly how this component behaves.

`modCursor()` in spring rest event may also cause other dilemmas. Say which cursor should we pass to the `props.onRest()`, the cursor before moding or after moding?

xiaody commented on src/TouchCarousel.js:
```diff
@@ -134,9 +160,21 @@ class TouchCarousel extends React.PureComponent {
+    }
+    this.isMouseDown = false
+    this.onEndOrCancel(1, e)
+  }
```
Does desktop has something like the `touchCancel` event? What happens if the user switches the browser tab or minimizes the browser window during dragging?

omahlama replied:
Good point, the correct event is `blur`. I added it to the event listeners.

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,22 +55,37 @@ class TouchCarousel extends React.PureComponent {
 
   componentDidMount () {
+    document.addEventListener('mousemove', this.onDocumentMouseMove)
+    document.addEventListener('mouseup', this.onDocumentMouseUp)
```
It would be nice if we add a prop and make all the mouse support stuff opt-in. It's "touch-carousel" after all.

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,16 +55,35 @@ class TouchCarousel extends React.PureComponent {
 
+  mockTouchEvent = (e) => {
+    e.changedTouches = [{ identifier: this.mouseTouchId, pageX: e.pageX, pageY: e.pageY }]
+    return e
```
Does a `blur` event has sane pageX and pageY?

And I think we'd better not modify the original events. Returning a plain object is good enough.
 
Also, `e.type` matters. It is used to choose which callback we call, onDragEnd or onDragCancel. See `onTouchEndOrCancel()`

omahlama commented:
I use plain objects now in the mocked events, had to do a `.bind` hack to get preventDefault working

xiaody commented:
Looks good to me. Really appreciate the work. 👍

Would you like squash the commits manually or let me use the "squash and merge" that github provides? I've never used the later one and I'm not sure what it does the to commit author field.

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,16 +56,59 @@ class TouchCarousel extends React.PureComponent {
+  addMouseListeners () {
+    document.addEventListener('mousemove', this.onDocumentMouseMove)
+    document.addEventListener('mouseup', this.onDocumentMouseUp)
+    document.addEventListener('blur', this.onDocumentMouseUp)
```
It seems `blur` happens on window, not document.

xiaody commented on README.md:
```diff
@@ -131,6 +131,10 @@ If `deltCrossAxis * ignoreCrossMove > deltMainAxis`, carousel would ignore the d
 
+### props.mouseSupport {Boolean}
+
+Set this is set to `true`, desktop users can drag the carousel using the mouse.
```
s/Set/If/

xiaody commented on src/TouchCarousel.js:
```diff
@@ -55,16 +56,59 @@ class TouchCarousel extends React.PureComponent {
+
+  mockTouchEvent = (e) => ({
+    changedTouches: [{ identifier: this.mouseTouchId, pageX: e.pageX, pageY: e.pageY }],
+    type: e.type === 'mouseup' ? 'touchend' : e.type,
```
More accurate:
mousedown -> touchstart
mousemove -> touchmove
mouseup -> touchend
blur -> touchcancel

omahlama commented:
I can squash them, I’ll do it after lunch.

Olli

> xiaody <notifications@github.com> kirjoitti 14.12.2017 kello 11.50:
> 
> Looks good to me. Really appreciate the work. 👍
> 
> Would you like squash the commits manually or let me use the "squash and merge" that github provides? I've never used the later one and I'm not sure what it does the to commit author field.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

omahlama commented:
I fixed the cr comments. You were right about window.blur, once it actually started working the missing pageX caused issues - I fixed by adding a check and going to the "snap to closest" branch of the if.

omahlama commented:
Now the commits are squashed into one

xiaody commented:
Great. Will do the merge and publish tomorrow.

It strikes me that since all the mouse support code are clearly separate and built on top of touch support, maybe we can move the code to a separate file and provide a high order component. Something like:

```jsx
import withMouseSupport from 'react-touch-carousel/lib/mouseSupportHOC'

const Container = withMouseSupport(CarouselContainer)

<TouchCarousel
  component={Container}
/>
```

Does it sounds reasonable for you? I can try to do this at this weekend if you don't have enough time.

omahlama commented:
That sounds very good, keeps the code size smaller for people not using mouse events. I really should get back to building features on top of touch-carousel, so if you have time for this refactoring it'd be great.

xiaody commented:
Your effort is never wasted. I move the code a bit and realize that the logic is so common that it's even worth a separate package.

```jsx
/* react-touch-with-mouse prototype */
import React from 'react'

export default function touchWithMouseHOC (Component) {
  class TouchWithMouse extends React.Component {
    constructor (props) {
      super(props)
      this.isMouseDown = false
      this.mouseDownId = 0
    }

    componentDidMount () {
      document.addEventListener('mousemove', this.onDocumentMouseMove)
      document.addEventListener('mouseup', this.onDocumentMouseUp)
      window.addEventListener('blur', this.onDocumentMouseUp)
    }

    componentWillUnmount () {
      document.removeEventListener('mousemove', this.onDocumentMouseMove)
      document.removeEventListener('mouseup', this.onDocumentMouseUp)
      window.removeEventListener('blur', this.onDocumentMouseUp)
    }

    mockTouchEvent () {}

    onMouseDown = (e) => {
      this.props.onMouseDown(e)
      this.isMouseDown = true
      this.mouseDownId++
      this.props.onTouchStart(this.mockTouchEvent(e))
    }

    onDocumentMouseMove () {}

    onDocumentMouseUp () {}

    render () {
      const {onMouseDown, ...rest} = this.props
      return (
        <Component
          {...rest}
          onMouseDown={this.onMouseDown}
        />
      )
    }
  }

  TouchWithMouse.defaultProps = {
    onMouseDown () {},
    onTouchStart () {},
    onTouchMove () {},
    onTouchEnd () {},
    onTouchCancel () {}
  }

  TouchWithMouse.displayName = `TouchWithMouse(${Component.displayName || Component.name})`

  return TouchWithMouse
}
```

Isn't it? So now I encourage you to publish a standalone package. And when you are done, I will update touch-carousel's README to refer your new package.

omahlama commented:
I would keep this as part of the react-touch-carousel package, as the touch emulation is tightly coupled with the functionality of carousel: the fact that mousemove and mouseup events are attached to `document` and not to `Component` makes sense with the carousel, but not in a general case. Also, the mocked events contain only the fields used by the carousel, not everything specified by the spec (https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent). Developing a full touch emulation HOC would be much bigger task.

xiaody commented:
Moved the mouse support code to `touchWithMouseHOC.js` and released `react-touch-carousel@0.8.0`
